File: 73. Set Matrix Zeroes/73. Set Matrix Zeroes.rb
# @param {Integer[][]} matrix
# @return {Void} Do not return anything, modify matrix in-place instead.
def set_zeroes(matrix)
    m, n = matrix.length, matrix[0].length
    first_row, first_col = false, false
    
    # Check if first row should be set to zero
    n.times do |j|
        if matrix[0][j] == 0
            first_row = true
            break
        end
    end
    
    # Check if first column should be set to zero
    m.times do |i|
        if matrix[i][0] == 0
            first_col = true
            break
        end
    end
    
    # Use first row and column as markers
    (1...m).each do |i|
        (1...n).each do |j|
            if matrix[i][j] == 0
                matrix[i][0] = 0
                matrix[0][j] = 0
            end
        end
    end
    
    # Set zeros based on markers
    (1...m).each do |i|
        (1...n).each do |j|
            if matrix[i][0] == 0 || matrix[0][j] == 0
                matrix[i][j] = 0
            end
        end
    end
    
    # Set first row to zero if needed
    if first_row
        n.times do |j|
            matrix[0][j] = 0
        end
    end
    
    # Set first column to zero if needed
    if first_col
        m.times do |i|
            matrix[i][0] = 0
        end
    end
end

File: 73. Set Matrix Zeroes/73. Set Matrix Zeroes.c
void setZeroes(int** matrix, int matrixSize, int* matrixColSize) {
    int m = matrixSize, n = matrixColSize[0];
    int firstRow = 0, firstCol = 0;
    
    // Check if first row should be set to zero
    for (int j = 0; j < n; j++) {
        if (matrix[0][j] == 0) {
            firstRow = 1;
            break;
        }
    }
    
    // Check if first column should be set to zero
    for (int i = 0; i < m; i++) {
        if (matrix[i][0] == 0) {
            firstCol = 1;
            break;
        }
    }
    
    // Use first row and column as markers
    for (int i = 1; i < m; i++) {
        for (int j = 1; j < n; j++) {
            if (matrix[i][j] == 0) {
                matrix[i][0] = 0;
                matrix[0][j] = 0;
            }
        }
    }
    
    // Set zeros based on markers
    for (int i = 1; i < m; i++) {
        for (int j = 1; j < n; j++) {
            if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                matrix[i][j] = 0;
            }
        }
    }
    
    // Set first row to zero if needed
    if (firstRow) {
        for (int j = 0; j < n; j++) {
            matrix[0][j] = 0;
        }
    }
    
    // Set first column to zero if needed
    if (firstCol) {
        for (int i = 0; i < m; i++) {
            matrix[i][0] = 0;
        }
    }
}